exla/lib/exla/defn.ex

```diff
@@ -466,7 +467,9 @@ defmodule EXLA.Defn do
                   )
                 end)
 
-              {:ok, {xla_time, executable, inputs_and_typespecs, %{outfeed | infeeds: []}}}
+              {:ok,
+               {xla_time, executable, {inputs_and_typespecs, used_inputs, outputs},
+                %{outfeed | infeeds: []}}}
```

Comment:
Adding `outputs` and `used_inputs` here means those values will be cached twice. I think for `to_mlir_module`, we don't want to use the cache funs anymore. I would instead check inside `__compile__` if `module_compilation` is set to `:mlir` and have it throw there the exact data that you need.

Alternatively you can create a new EXLA.Defn.MLIRCompiler, that calls `EXLA.Defn`. We would need to make some APIs public.